File: Sheets/src/neetcode/practice/Trees/_1_invertBinaryTree.java
package neetcode.practice.Trees;

class Solution {
  public TreeNode invertTree(TreeNode root) {
    Procedure proc = node -> {
      TreeNode temp = node.left;
      node.left = node.right;
      node.right = temp;
    };
    this.dfs(root, proc);
    return root;
  }

  private void dfs(TreeNode root, Procedure proc) {
    if (root == null)
      return;

    proc.run(root);

    this.dfs(root.left, proc);
    this.dfs(root.right, proc);
  }
}

interface Procedure {
  public void run(TreeNode root);
}

File: Sheets/src/neetcode/practice/Trees/_1_InvertBinaryTree.cc
#include "./TreeNode.hpp"

class Solution {
public:
  void dfs(TreeNode *root) {
    if (root == nullptr)
      return;

    TreeNode *temp = root->left;
    root->left = root->right;
    root->right = temp;

    dfs(root->left);
    dfs(root->right);
  }

  TreeNode *invertTree(TreeNode *root) {
    dfs(root);
    return root;
  }
};
